### PR TITLE
Jetpack Assistant: Add client side state

### PR DESCRIPTION
**Pull request by robertf4**
#### Changes proposed in this Pull Request:

This PR adds the client side state for the new Jetpack Assistant.

Includes:
* Actions and reducers to support all the state needs of the Assistant.
* Updates to the initial-state reducer.
* Updates to rest-api and corresponding data components for loading the Assistant state from the server.
* Some small changes to the plans constants and the site reducer.

#### Testing instructions:

Functional testing with this PR directly isn't possible without more client side functionality, so testing will have to wait for a later PR.

#### Proposed changelog entry for your changes:

* General: Add new Jetpack Recommendations wizard to help you get started with our recommended features of Jetpack.

### CODE REVIEW

File: _inc/client/components/data/query-assistant-data/index.jsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+
+/**
+ * Internal dependencies
+ */
+import { fetchRecommendationsData, isFetchingRecommendationsData } from 'state/recommendations';
+import { isOfflineMode } from 'state/connection';
+
+class QueryRecommendationsData extends Component {
```
Comment (brbrr) [inline _inc/client/components/data/query-assistant-data/index.jsx:14]: This file looks identical to `query-recommendations-data/index.jsx`. Is it expected?

File: _inc/client/state/recommendations/actions.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+		.saveRecommendationsData( recommendations.data )
+		.then( () => {
+			dispatch( { type: JETPACK_RECOMMENDATIONS_DATA_SAVE_SUCCESS } );
+		} )
+		.catch( error => {
+			dispatch( { type: JETPACK_RECOMMENDATIONS_DATA_SAVE_FAIL, error } );
+		} );
+};
+
+export const saveRecommendationsData = () => {
```
Comment (brbrr) [inline _inc/client/state/recommendations/actions.js:56]: It looks like this function is not used anywhere. Do we still need it?

File: _inc/client/state/recommendations/reducer.js [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+	JETPACK_RECOMMENDATIONS_UPSELL_FETCH,
+	JETPACK_RECOMMENDATIONS_UPSELL_FETCH_RECEIVE,
+	JETPACK_RECOMMENDATIONS_UPSELL_FETCH_FAIL,
+} from 'state/action-types';
+import { getRewindStatus } from 'state/rewind';
+import { getSetting } from 'state/settings';
+import { getSitePlan, hasActiveProductPurchase, hasActiveScanPurchase } from 'state/site';
+import { isPluginActive } from 'state/site/plugins';
+
+const mergeArrays = ( x, y ) => {
```
Comment (brbrr) [inline _inc/client/state/recommendations/reducer.js:30]: Why do we need this function? Is `mergeWith` not enough?

File: _inc/client/components/data/query-assistant-data/index.jsx (hunk 1 above)
Comment (robertf4) [reply _inc/client/components/data/query-assistant-data/index.jsx]: I missed this one during a refactor, I've deleted it now.

File: _inc/client/state/recommendations/actions.js (hunk 2 above)
Comment (robertf4) [reply _inc/client/state/recommendations/actions.js:56]: It is used in the UI which is in a different PR, see [here](https://github.com/Automattic/jetpack/pull/17691/files#diff-78e1e217c7c497d7595da78066235a28807c6696269605495bcc84ba6af35149R111) for an example.

File: _inc/client/state/recommendations/reducer.js (hunk 3 above)
Comment (robertf4) [reply _inc/client/state/recommendations/reducer.js:30]: `merge` on its own doesn't merge the array properties like I need it to here, so I used `mergeWith` and this customizer function to get the desired behavior. See this jsfiddle for an illustration of the differences between `merge` and `mergeWith`: https://jsfiddle.net/8jzfc0qt/.

**Review by jeherve — APPROVED** [review]
This looks good to me. ✔️

**Review by jeherve — DISMISSED** [review]
Re-approving after a rebase to solve unrelated test failures.

**Comment by robertf4** [comment]
Fixed some ESLint errors that were preventing me from merging.

**Review by anomiex — CHANGES_REQUESTED** [review]
You should fix the warnings too. They're supposed to be making CI fail too, but due to an oversight they aren't currently.

**Review by kraftbj — APPROVED** [review]

**Review by anomiex — APPROVED** [review]
Jeremy said it looked good earlier, and the changes since to pass CI look good to me.

**Pull request merged**